File: leetcode/LeetCode_75/merge-strings-alternately/ruby/solution.rb
# echo ./solution.rb | entr -c ruby /_

def pad(arr, final_len)
  pad_len = final_len - arr.length
  if pad_len > 0 then
    arr.concat([nil] * pad_len)
  else
    arr
  end
end

def interleaved(word1, word2)
  largest = [word1.length, word2.length].max
  chars1 = pad(word1.chars, largest)
  chars2 = pad(word2.chars, largest)

  chars1.zip(chars2)
    .flatten
    .compact
    .join
end


require 'minitest/autorun'
class TestSolution < Minitest::Test

  def test_padding
   assert_equal([1,2,3,nil,nil], pad([1,2,3], 5))
   assert_equal([1,2,3], pad([1,2,3], 3))
   assert_equal([1,2,3], pad([1,2,3], 0))
   assert_equal([], pad([], 0))
   assert_equal([nil,nil,nil], pad([], 3))
  end

  def test_solution
    assert_equal("ABCDEF", interleaved("AC", "BDEF"))
    assert_equal("ABCDEF", interleaved("ACEF", "BD"))
    assert_equal("ABCD", interleaved("AC", "BD"))
    assert_equal("", interleaved("", ""))
  end

end

File: leetcode/LeetCode_75/merge-strings-alternately/elixir/solution.ex
# echo ./solution.ex | entr -c elixir /_

ExUnit.start()

defmodule Solution do
  def interleave_letters("", ""), do: ""
  def interleave_letters(<<x::binary-size(1)>> <> xs, ""), do: x <> interleave_letters(xs, "")
  def interleave_letters("", <<y::binary-size(1)>> <> ys), do: y <> interleave_letters("", ys)

  def interleave_letters(<<x::binary-size(1)>> <> xs, <<y::binary-size(1)>> <> ys) do
    x <> y <> interleave_letters(xs, ys)
  end
end

defmodule Test do
  use ExUnit.Case, async: true

  test "interleave_letters" do
    assert "ABCD" == Solution.interleave_letters("AC", "BD")
    assert "ABCD_EF" == Solution.interleave_letters("AC", "BD_EF")
    assert "ABCD_EF" == Solution.interleave_letters("AC_EF", "BD")
    assert "ABC" == Solution.interleave_letters("ABC", "")
    assert "ABC" == Solution.interleave_letters("", "ABC")
    assert "" == Solution.interleave_letters("", "")
  end
end
